tokio-util/src/future.rs

```diff
@@ -0,0 +1,124 @@
+//! An extension trait for Futures that provides a variety of convenient adapters.
+
+mod with_cancellation_token;
+use with_cancellation_token::{WithCancellationTokenFuture, WithCancellationTokenFutureOwned};
+
+use std::future::Future;
+
+use crate::sync::CancellationToken;
+
+/// A trait which contains a variety of convenient adapters and utilities for `Future`s.
+pub trait FutureExt: Future {
+    /// A wrapper around [`tokio::time::timeout`], with the advantage that it is easier to write
+    /// fluent call chains.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use tokio::{sync::oneshot, time::Duration};
+    /// use tokio_util::future::FutureExt;
+    ///
+    /// # async fn dox() {
+    /// let (tx, rx) = oneshot::channel::<()>();
+    ///
+    /// let res = rx.timeout(Duration::from_millis(10)).await;
+    /// assert!(res.is_err());
+    /// # }
+    /// ```
+    #[cfg(feature = "time")]
+    fn timeout(self, timeout: std::time::Duration) -> tokio::time::Timeout<Self>
+    where
+        Self: Sized,
+    {
+        tokio::time::timeout(timeout, self)
+    }
+
+    /// A wrapper around [`tokio::time::timeout_at`], with the advantage that it is easier to write
+    /// fluent call chains.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use tokio::{sync::oneshot, time::{Duration, Instant}};
+    /// use tokio_util::future::FutureExt;
+    ///
+    /// # async fn dox() {
+    /// let (tx, rx) = oneshot::channel::<()>();
+    /// let deadline = Instant::now() + Duration::from_millis(10);
+    ///
+    /// let res = rx.timeout_at(deadline).await;
+    /// assert!(res.is_err());
+    /// # }
+    /// ```
+    #[cfg(feature = "time")]
+    fn timeout_at(self, deadline: tokio::time::Instant) -> tokio::time::Timeout<Self>
+    where
+        Self: Sized,
+    {
+        tokio::time::timeout_at(deadline, self)
+    }
+
+    /// Similar to [`CancellationToken::run_until_cancelled`], but with the advantage that it is easier to write
+    /// fluent call chains, and biased towards waiting for [`CancellationToken`] to complete.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use tokio::sync::oneshot;
+    /// use tokio_util::future::FutureExt;
+    /// use tokio_util::sync::CancellationToken;
+    ///
+    /// # async fn dox() {
+    /// let (tx, rx) = oneshot::channel::<()>();
+    /// let token = CancellationToken::new();
+    /// let token_clone = token.clone();
+    /// tokio::spawn(async move {
+    ///     tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+    ///     token.cancel();
+    /// });
+    /// assert!(rx.with_cancellation_token(&token_clone).await.is_none())
+    /// # }
+    /// ```
+    fn with_cancellation_token(
+        self,
+        cancellation_token: &CancellationToken,
+    ) -> WithCancellationTokenFuture<'_, Self>
```

Comment:
Please also highlight the fairness here, we could refer the below one.

https://github.com/tokio-rs/tokio/blob/911ab21d7012a50e53971ad1292a9f18de22d4c8/tokio-util/src/sync/cancellation_token.rs#L271-L275